### PR TITLE
[UI Tests] Fix flaky `allDayStatsLoad` test

### PR DESCRIPTION
**Pull request by pachlava**
### Edit

After the initial PR description was created 3 weeks ago, the PR was blocked from merging because of the [java.lang.OutOfMemoryError: Java heap space](https://github.com/wordpress-mobile/WordPress-Android/pull/16838#issuecomment-1171133746) error. The error was solved after mirgating from CircleCI to Buildkite, however, the test started to fail in a new way (possibly because of app using Compose), which was adressed in d7576688ec9204c448a07edcedfb1f23a4c6a627.

---------------------------
### Why

`allDayStatsLoad` UI test continued to fail after #16799 merge.

### How

I took a look into Firebase videos and could not see the test being even executed. After that, I tried the local execution and finally saw the fail: it appears that `allDayStatsLoad` is executed after `signUpWithMagicLink` test. The latter test left the app in a weird state, my guess is because it emulated using the magic link signup, which probably did not work well with WireMock. The app then crashed after relaunch and opening Settings (to log out). So, in fact, the reason for the fail was in the other test and had nothing to do with `allDayStatsLoad`.

<details>
	<summary>Crash details</summary>

```
2022-06-29 22:59:58.355 15086-15086/org.wordpress.android.prealpha E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.wordpress.android.prealpha, PID: 15086
    java.lang.IllegalStateException: Settings should be hidden when isSettingsSupported returns false.
        at org.wordpress.android.ui.reader.viewmodels.ReaderViewModel.onSettingsActionClicked(ReaderViewModel.kt:223)
        at org.wordpress.android.ui.reader.ReaderFragment.onCreateOptionsMenu$lambda-4$lambda-3(ReaderFragment.kt:120)
        at org.wordpress.android.ui.reader.ReaderFragment.$r8$lambda$rHyWAw1DWBeB0Lu-hW7SiIIOon0(Unknown Source:0)
        at org.wordpress.android.ui.reader.ReaderFragment$$ExternalSyntheticLambda1.onClick(Unknown Source:2)
        at android.view.View.performClick(View.java:6597)
        at android.view.View.performClickInternal(View.java:6574)
        at android.view.View.access$3100(View.java:778)
        at android.view.View$PerformClick.run(View.java:25885)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

</details>

I've added a teardown with logout to `signUpWithMagicLink`, which is done while the app was still running after using the magic link, which seems to solve the problem.

### To test

- Optionally, try running the whole package of `e2e` test locally
- CI is green for the last 7 executions (check the commits history) or see the Firebase logs (just me being super paraniod, so I masked the executions IDs 😅) :

Screenshot 2022-06-30 at 10 37 47

### CODE REVIEW

**Comment by ParaskP7** [comment]
👋 @pachlava !

About this `buildkite/wordpress-android/test-processors` check that is expected and waiting for status to be reported, it wouldn't work unless you update this PR related `UI-tests-examination` branch of yours with the latest `trunk` and push the changes. Can you do that? 🙇

**Review by ParaskP7 — APPROVED** [review]
👋 @pachlava !

I have reviewed this change and it LGTM, thank you for looking into this UI test related problem and trying to fix this `allDayStatsLoad` flaky test of ours! 💯 🙇 🥇 

> I've added a teardown with logout to signUpWithMagicLink, which is done while the app was still running after using the magic link, which seems to solve the problem.

Nice investigation there! 🔍

It does seems that this might do the trick. I am also seeing that the `LoginTests` test suite have this `logoutIfNecessary()` logic, both during `setUp()` and `tearDown()`. Thus, it seems logical to me that the `SignUpTests` test suite might benefit from it as well. 🤞

**Comment by pachlava** [comment]
Thank you for the review, @ParaskP7 👋 !

> About this buildkite/wordpress-android/test-processors check that is expected and waiting for status to be reported, it wouldn't work unless you update this PR related UI-tests-examination branch of yours with the latest trunk and push the changes. Can you do that? 🙇

I just did, and it did help in this respect. But the build started to fail for me, in the same way it does on `trunk` since recently 😅 

I guess I need to wait for the fix before merging.

**Comment by ParaskP7** [comment]
👋 @pachlava and thank you for the update! 🙇 

> I guess I need to wait for the fix before merging.

Yea, I am also seeing this `Exception in thread "Daemon client event forwarder" java.lang.OutOfMemoryError: Java heap space` and I am not sure why it started happening... 🤔 

I am shamelessly cc'ing @oguzkocer and @jkmassel , maybe you know more about this? 🙏

**Comment by oguzkocer** [comment]
@pachlava @ParaskP7 You can replace [these lines](https://github.com/wordpress-mobile/WordPress-Android/blob/UI-tests-examination/.circleci/config.yml#L153-L154) with just `- update-gradle-memory` (notice the removal of `:` from the end) so that it'll increase the available memory for this task to the default `2048m` which _might_ fix the issue.

If that's not enough, we can try increasing the [default memory setting for CircleCI](https://github.com/wordpress-mobile/WordPress-Android/blob/UI-tests-examination/.circleci/config.yml#L26) up to `4G`. I _think_ our plan only has support for `2G` but the `Executor / Resource Class` in the [CircleCI job](https://app.circleci.com/pipelines/github/wordpress-mobile/WordPress-Android/32307/workflows/1ca2a386-639b-4354-937d-2d694bcca22e/jobs/125701) shows `4G` available memory, so it's worth trying it.

If that doesn't work either, I am afraid it'll have to wait for the Buildkite migration. The good news is I think we are almost ready for it, the bad news is we have been in that state for a while now, so 😓 🤷

**Comment by pachlava** [comment]
Thanks, @oguzkocer! Unfortunately, both of the proposed changes ([one](b5e40aa22f6e5c403bdbaf61988f9af669566a23), [two](c368838929714c6b05a4ab740a0dbaedb73bbc4c)) did not help, so it looks like waiting for Buildkite is the remaining option 🥲

File: WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/StatsPage.java
Comment (pachlava) [inline WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/StatsPage.java:41]: I decided to add the explanation for this, because it's quite hard to guess otherwise, especially when using API < 29, which will not show the extra screens in the Layout Inspector (but they will still exist).

File: WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/StatsPage.java
Comment (pachlava) [inline WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/StatsPage.java:50]: The order does not affect anything, it just seemed logical to list the parent first.

File: WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/StatsPage.java
Comment (pachlava) [inline WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/StatsPage.java:45]: This is the "main" part of the fix - using this visible `coordinator_layout` as a parent for all stats cards, because otherwise Android Studio might locate the out-of-screen card templates, that have the same headers text, size, relative coordinates, etc. as the ones we need.

**Comment by pachlava** [comment]
@ParaskP7 👋 🙂 

I guess the PR is ready for re-review. Besides the details I've shared in Slack, [this comment](https://github.com/wordpress-mobile/WordPress-Android/pull/16838#discussion_r926890898) refers to the most recent cause for the fail, which was addressed in d7576688ec9204c448a07edcedfb1f23a4c6a627.

You might see another test - `publishFullPost` - failing in the history of executions. I'd like to keep it separate, because the fail is not related to the current PR, and work on it next. That test likely fails because of performance reasons: uploading a featured image is one of the steps there.

File: WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/StatsPage.java
Comment (ParaskP7) [reply WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/StatsPage.java:41]: Great, thanks for adding this bit of extra explanation here @pachlava ! 🥇

**Comment by ParaskP7** [comment]
👋 @pachlava !

> I guess the PR is ready for re-review. Besides the details I've shared in Slack, https://github.com/wordpress-mobile/WordPress-Android/pull/16838#discussion_r926890898 refers to the most recent cause for the fail, which was addressed in https://github.com/wordpress-mobile/WordPress-Android/commit/d7576688ec9204c448a07edcedfb1f23a4c6a627.

Great, thanks for adding all these details here for me to know were to focus my review! 🥇 

PS: Please expect me to review this today, but after I am done with my release management tasks for today. 👍 

> You might see another test - publishFullPost - failing in the history of executions. I'd like to keep it separate, because the fail is not related to the current PR, and work on it next. That test likely fails because of performance reasons: uploading a featured image is one of the steps there.

I agree, let's work on `publishFullPost` separately and next in line, thank YOU! 💯

**Comment by pachlava** [comment]
> Please expect me to review this today, but after I am done with my release management tasks for today. 👍

Sure, no worries, and thank you, @ParaskP7!

**Review by ParaskP7 — APPROVED** [review]
👋 @pachlava !

I reviewed the latest changes to this PR and it all LGTM, thanks for working on it! 🥇 🙇 🚀 

Since the [instrumented-tests](https://buildkite.com/automattic/wordpress-android/builds/5642#01822195-c203-4407-a2b8-32cedbb02749) check on Buildkite is passing and the [allDayStatsLoad](https://console.firebase.google.com/project/api-project-108380595987/testlab/histories/bh.b3d1e88d8b09c8be/matrices/5675846366540032785/executions/bs.a4403e6e1a77660f/testcases/28/test-cases) test is 🟢 in Firebase Test Lab, I am approving this PR as I am confident in merging this to `trunk`. Let's now hope it's not going to end-up being flaky afterwards... 😅 🤞

Once more, great work on this Sergiy! ❤️ 💯 🌟

**Pull request merged**
